third_party/xla_client/pjrt_computation_client.h

```diff
@@ -0,0 +1,157 @@
+#ifndef XLA_CLIENT_PJRT_COMPUTATION_CLIENT_H_
+#define XLA_CLIENT_PJRT_COMPUTATION_CLIENT_H_
+
+#include <cstdint>
+
+#include "absl/types/span.h"
+#include "tensorflow/compiler/xla/client/xla_computation.h"
+#include "tensorflow/compiler/xla/literal.h"
+#include "tensorflow/compiler/xla/pjrt/pjrt_client.h"
+#include "tensorflow/compiler/xla/shape.h"
+#include "tensorflow/compiler/xla/xla_client/computation_client.h"
+#include "tensorflow/compiler/xla/xla_client/debug_macros.h"
+#include "tensorflow/compiler/xla/xla_client/util.h"
+
+namespace xla {
+
+class PjRtComputationClient : public ComputationClient {
+  struct PjRtData : public Data {
+    PjRtData(std::string device, Shape device_shape)
+        : Data(std::move(device), std::move(device_shape)) {}
+
+    PjRtData(std::string device, Shape device_shape,
+             std::shared_ptr<PjRtBuffer> buffer)
+        : Data(std::move(device), std::move(device_shape)), buffer(buffer) {}
+
+    void* get_handle() const {
+      return buffer->AcquireExternalReference()
+          .ValueOrDie()
+          ->OpaqueDeviceMemoryDataPointer();
+    };
+    OpaqueHandle GetOpaqueHandle() override {
+      return reinterpret_cast<std::uintptr_t>(get_handle());
+    };
+    void Assign(const Data& data) override;
+    bool HasValue() const override { return get_handle() != nullptr; };
+
+    std::shared_ptr<PjRtBuffer> buffer;
+  };
+
+  struct PjRtComputation : public Computation {
+    PjRtComputation(PjRtClient* self, XlaComputation computation,
+                    ProgramShape program_shape,
+                    std::vector<std::string> devices,
+                    xla::CompileOptions options)
+        : Computation(std::move(computation), std::move(program_shape),
+                      std::move(devices)),
+          executable(self->Compile(this->computation(), options).ValueOrDie()) {
```

Comment:
what's the motivation here of moving the compilation to the `PjRtComputation` constructor? I felt like leaving it out and explicitly taking a  `executable` might be more obvious. In the Xrt world, computation is return as the result of the `Compile` call.

Reply:
maybe, I saw in the Xrt Compile it took the `vecvtor<CompileInstance>` and it is passed by value. Worse case we can also pass `CompileInstance` by value. It would also be helpful to get the stack trace when it segfault so we know the exact issue.

Reply:
I made this constructor less clunky as you suggested and am not getting an error now. There must have been a mistake somewhere else.